storage/shard.go

```diff
@@ -990,31 +952,64 @@ func (s *dbShard) tryRetrieveWritableSeries(id ident.ID) (
 	return nil, opts, nil
 }
 
-func (s *dbShard) newShardEntry(id ident.ID, tags ident.TagIterator) (*dbShardEntry, error) {
-	clonedTags, err := s.cloneTags(tags)
-	if err != nil {
-		return nil, err
-	}
+func (s *dbShard) newShardEntry(
+	id ident.ID,
+	tags ident.TagIterator,
+) (*lookup.Entry, error) {
 	series := s.seriesPool.Get()
-	clonedID := s.identifierPool.Clone(id)
-	series.Reset(clonedID, clonedTags, s.seriesBlockRetriever,
-		s.seriesOnRetrieveBlock, s, s.seriesOpts)
-	uniqueIndex := s.increasingIndex.nextIndex()
-	return &dbShardEntry{series: series, index: uniqueIndex}, nil
-}
-
-func (s *dbShard) cloneTags(tags ident.TagIterator) (ident.Tags, error) {
+	// NB(r): As documented in storage/series.DatabaseSeries the series IDs
+	// are garbage collected, hence we cast the ID to a BytesID that can't be
+	// finalized.
+	// Since series are purged so infrequently the overhead of not releasing
+	// back an ID to a pool is amortized over a long period of time.
+	clonedID := ident.BytesID(append([]byte(nil), id.Bytes()...))
+
+	// Inlining tag creation here so its obvious why we can safely index
+	// into clonedID below
 	tags = tags.Duplicate()
-	clone := s.identifierPool.Tags()
-	defer tags.Close()
+	clonedTags := s.identifierPool.Tags()
+
+	// Avoid finalizing the tags since series will let them be garbage collected
+	clonedTags.NoFinalize()
+
 	for tags.Next() {
 		t := tags.Current()
-		clone.Append(s.identifierPool.CloneTag(t))
+
+		// NB(r): Optimization for workloads that embed the tags in the ID is to
```

Comment:
I still can't decide if this is brilliant or terrible.

Reply:
Heheh, talked with Richie about it and agreed we should flag this with config but default it to on.

Reply:
I'll add this.